File: AnalysisCode/di_scrubbing_func.m
function [plot_dat, OPGEE_bin] = di_scrubbing_func(M_raw, Basin_Select, Basin_Index, activityfolder)


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% This function was adjusted based on the code "di_scrubbing_alldata.m"
%   C:\Users\jruthe\Dropbox\Doctoral\Projects\Research Projects\manuscripts\2_BU_paper\3_Analysis\4_OPGEE_Modelling\a_Scenario_Generation\5b_DI Processing\REVISED GITHUB
%
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% Parameters for processing

    cutoff = 100;% Mscf/bbl

% Data is organized as follows, for BOTH basin level data and the 2015 US-level data from David lyon:

% Col 1 = oil (bbl/year)
% Col 2 = gas (Mscf/year)
% Col 3 = well count (in TX, a row is a lease with multiple wells, source =
% David Lyons)

%% Array Transformation

% First transform the array so that each row corresponds to a well

[size_mat,~] = size(M_raw);
M_raw(:,3) = round(M_raw(:,3));
total_wells = sum(M_raw(:,3));
M_new = zeros(total_wells,4);


row = 1;
for i = 1:size_mat
    wells = M_raw(i,3);
    if wells == 1
        M_new(row,1) = M_raw(i,1)/365.25;
        M_new(row,2) = M_raw(i,2)/365.25;
        row = row + 1;
    else
        for j = 1:wells
            prod.oil = M_raw(i,1)/wells/365.25; % Convert to bbl/day
            prod.gas = M_raw(i,2)/wells/365.25; % Convert to Mscf/day
            M_new(row,1) = prod.oil;
            M_new(row,2) = prod.gas;
            row = row + 1;
        end
    end

end

plot_dat = M_new(:,2);
[M_no_offshore,count,totalprod,averageprod] = data_class(M_new(:,1:2), cutoff, activityfolder);

%     %   PRINT OUTPUTS FROM THIS SECTION
%     
%         % Header
%         fprintf(1,'        # wells     Total oil (MMbbls)     Total gas (Bscf/yr)     Average oil (bbl/day well)     Average gas (Mscf/day well)\n');
%         % Data
%         fprintf(1,'G only  %0.0f        0                 %0.0f             0                           %4.2f\n' ,count.drygas,totalprod.drygas(1,2),averageprod.drygas(1,2));
%         fprintf(1,'G w oil %0.0f        %0.0f             %0.0f             %4.2f                       %4.2f\n' ,count.gaswoil,totalprod.gaswoil(1,1),totalprod.gaswoil(1,2),averageprod.gaswoil(1,1),averageprod.gaswoil(1,2));
%         fprintf(1,'O only  %0.0f        %0.0f             0                 %4.2f                       0\n'     ,count.oil,totalprod.oil(1,1),averageprod.oil(1,1));
%         fprintf(1,'O w gas %0.0f        %0.0f             %0.0f             %4.2f                       %4.2f\n' ,count.oilwgas,totalprod.oilwgas(1,1),totalprod.oilwgas(1,2),averageprod.oilwgas(1,1),averageprod.oilwgas(1,2));
%         fprintf(1,'G total %0.0f        %0.0f             %0.0f             %4.2f                       %4.2f\n' ,count.gasall,totalprod.gasall(1,1),totalprod.gasall(1,2),averageprod.gasall(1,1),averageprod.gasall(1,2));   
%         fprintf(1,'O only+ %0.0f        %0.0f             %0.0f             %4.2f                       %4.2f\n' ,count.oil,totalprod.oil(1,1),totalprod.oil(1,2),averageprod.oil(1,1),averageprod.oil(1,2));
%         fprintf(1,'O total %0.0f        %0.0f             %0.0f             %4.2f                       %4.2f\n' ,count.oilall,totalprod.oilall(1,1),totalprod.oilall(1,2),averageprod.oilall(1,1),averageprod.oilall(1,2));

data_tab = cell(3,4);

data_tab(1,2) = cellstr('# wells');
data_tab(1,3) = cellstr('Total oil (MMbbls)');
data_tab(1,4) = cellstr('Total gas (Bscf/yr)');
data_tab(2,1) = cellstr('Gas wells');
data_tab(2,2) = num2cell(count.gasall);           %# wells
data_tab(2,3) = num2cell(totalprod.gasall(1,1));  %Total oil (MMbbls)
data_tab(2,4) = num2cell(totalprod.gasall(1,2));  %Total gas (Bscf/yr)
data_tab(3,1) = cellstr('Oil wells');
data_tab(3,2) = num2cell(count.oilall);           %# wells
data_tab(3,3) = num2cell(totalprod.oilall(1,1));  %Total oil (MMbbls)
data_tab(3,4) = num2cell(totalprod.oilall(1,2));  %Total gas (Bscf/yr)        

%cd 'C:\Users\jruthe\Dropbox\Doctoral\Projects\Research Projects\OPGEE\0_OPGEE_Matlab\Version 2\Outputs'   
if Basin_Select == 0
    FileName = 'DI_summary_US.xlsx';
else
    FileName = ['DI_summary_' Basin_Index{Basin_Select} 'out.xlsx'];
    filepath = fullfile(pwd, 'Outputs/',FileName);
end
xlswrite(filepath, data_tab)
%cd 'C:\Users\jruthe\Dropbox\Doctoral\Projects\Research Projects\OPGEE\0_OPGEE_Matlab\Version 2\'
 
%% BINNING - DRY GAS WELLS
        
%(i) Determine bin indices and bin means
    edges_gas_set = [0, 1, 5, 10, 20, 50, 100, 500, 1000, 10000,500000000];
    [counts_gas, edges_gas, ind_gas] = histcounts(M_no_offshore.drygas(:,2), edges_gas_set);
    
%(ii) Determine bin means
    bin_ave_gas = accumarray(ind_gas, M_no_offshore.drygas(:,2),[],@mean);
    bin_sum_gas = accumarray(ind_gas, M_no_offshore.drygas(:,2),[],@sum);
    bin_sum_oilwg = accumarray(ind_gas, M_no_offshore.drygas(:,1),[],@sum);

    bins_exp = zeros((length(edges_gas_set)-1),4);
    
    bins_exp(1:length(counts_gas'),1) = counts_gas';
    bins_exp(1:length(bin_ave_gas'),2) = bin_ave_gas';
    bins_exp(1:length(bin_sum_gas'),3) = bin_sum_gas';
    bins_exp(1:length(bin_sum_oilwg'),4) = bin_sum_oilwg';
    OPGEE_bin.gasdry = bins_exp;
%(iii) Print outputs to a text file
    
%       csvwrite(char(strcat(Basin_Index(1),'_gasdry_bins.csv')),bins_exp)  


%% BINNING - ASSOC GAS WELLS
      
%(i) Determine bin indices and bin means
    edges_gas_set = [0, 1, 5, 10, 20, 50, 100, 500, 1000, 10000,500000000];
    [counts_gas, edges_gas, ind_gas] = histcounts(M_no_offshore.gaswoil(:,2), edges_gas_set);
    
%(ii) Determine bin means
    bin_ave_gas = accumarray(ind_gas, M_no_offshore.gaswoil(:,2),[],@mean);
    bin_sum_gas = accumarray(ind_gas, M_no_offshore.gaswoil(:,2),[],@sum);
    bin_sum_oilwg = accumarray(ind_gas, M_no_offshore.gaswoil(:,1),[],@sum);

    bins_exp = zeros((length(edges_gas_set)-1),4);
    
    bins_exp(1:length(counts_gas'),1) = counts_gas';
    bins_exp(1:length(bin_ave_gas'),2) = bin_ave_gas';
    bins_exp(1:length(bin_sum_gas'),3) = bin_sum_gas';
    bins_exp(1:length(bin_sum_oilwg'),4) = bin_sum_oilwg';
    
%(iii) Print outputs to a text file
    
%       csvwrite(char(strcat(Basin_Index(1),'_gasassoc_bins.csv')),bins_exp)  
      OPGEE_bin.gasassoc = bins_exp;      
%% BINNING - OIL WELLS
        
%(ii) Determine bin indices and bin means

    edges_oil_set = [0, 1, 5, 10, 20, 50, 100, 500, 1000, 10000,1500000];
    [counts_oil, edges_oil, ind_oil] = histcounts(M_no_offshore.oilwgas(:,2), edges_oil_set);
    
% %(iii) Determine bin means

    bin_ave_gas = accumarray(ind_oil, M_no_offshore.oilwgas(:,2),[],@mean);
    bin_sum_gas = accumarray(ind_oil, M_no_offshore.oilwgas(:,2),[],@sum);
    bin_sum_oilwg = accumarray(ind_oil, M_no_offshore.oilwgas(:,1),[],@sum);

    bins_exp = zeros((length(edges_oil_set)-1),4);
    
    bins_exp(1:length(counts_oil'),1) = counts_oil';
    bins_exp(1:length(bin_ave_gas'),2) = bin_ave_gas';
    bins_exp(1:length(bin_sum_gas'),3) = bin_sum_gas';
    bins_exp(1:length(bin_sum_oilwg'),4) = bin_sum_oilwg';
    
%(iii) Print outputs to a text file
    
%       csvwrite(char(strcat(Basin_Index(1),'_oil_bins.csv')),bins_exp)
      OPGEE_bin.oilwgas = bins_exp; 
%% BINNING - WET WELLS

%(ii) Determine bin indices and bin means

    edges_oil_set = [0, 0.5, 1, 10,1500000];
    [counts_oil, edges_oil, ind_oil] = histcounts(M_no_offshore.oil(:,2), edges_oil_set);
    
% %(iii) Determine bin means

    bin_ave_gas = accumarray(ind_oil, M_no_offshore.oil(:,2),[],@mean);
    bin_sum_gas = accumarray(ind_oil, M_no_offshore.oil(:,2),[],@sum);
    bin_sum_oilwg = accumarray(ind_oil, M_no_offshore.oil(:,1),[],@sum);

    bins_exp = zeros((length(edges_oil_set)-1),4);
    
    bins_exp(1:length(counts_oil'),1) = counts_oil';
    bins_exp(1:length(bin_ave_gas'),2) = bin_ave_gas';
    bins_exp(1:length(bin_sum_gas'),3) = bin_sum_gas';
    bins_exp(1:length(bin_sum_oilwg'),4) = bin_sum_oilwg';
    
%(iii) Print outputs to a text file
    
%       csvwrite(char(strcat(Basin_Index(1),'_oilwet_bins.csv')),bins_exp)
      OPGEE_bin.oil = bins_exp; 
end

     

File: AnalysisCode_python/di_scrubbing_func.py
import numpy as np
import pandas as pd
import os
from data_class import data_class
import warnings
warnings.filterwarnings('ignore')

def di_scrubbing_func(M_raw, Basin_Select, Basin_Index, activityfolder):

    cutoff = 100  # Mscf/bbl
    M_raw[:, 2] = np.round(M_raw[:, 2])
    total_wells = int(np.sum(M_raw[:, 2]))

    # Initialize M_new
    M_new = np.zeros((total_wells, 4))

    # Transform the array so that each row corresponds to a well
    row = 0
    for i in range(M_raw.shape[0]):
        wells = int(M_raw[i, 2])
        if wells == 1:
            M_new[row, 0] = M_raw[i, 0] / 365.25
            M_new[row, 1] = M_raw[i, 1] / 365.25
            row += 1
        else:
            for _ in range(wells):
                prod_oil = M_raw[i, 0] / wells / 365.25  # Convert to bbl/day
                prod_gas = M_raw[i, 1] / wells / 365.25  # Convert to Mscf/day
                M_new[row, 0] = prod_oil
                M_new[row, 1] = prod_gas
                row += 1

    # Replace NaN values with 0
    M_new[np.isnan(M_new)] = 0
    plot_dat = M_new[:, 1]

    # Data classification
    M_no_offshore, count, totalprod, averageprod = data_class(M_new[:, :2], cutoff, activityfolder)

    # Create data table for gas and oil wells
    data_tab = [
        ["", "# wells", "Total oil (MMbbls)", "Total gas (Bscf/yr)"],
        ["Gas wells", count['gasall'], totalprod['gasall'][0], totalprod['gasall'][1]],
        ["Oil wells", count['oilall'], totalprod['oilall'][0], totalprod['oilall'][1]]
    ]

    # Save the data table to an Excel file
    if Basin_Select == 0:
        file_name = 'DI_summary_US.xlsx'
    else:
        file_name = f'DI_summary_{Basin_Index[Basin_Select]}out.xlsx'

    output_dir = 'Outputs'
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)
    filepath = os.path.join(output_dir, file_name)
    data_table = pd.DataFrame(data_tab)
    data_table.to_excel(filepath, header=False, index=False)

    # BINNING - DRY GAS WELLS
    edges_gas_set = [0, 1, 5, 10, 20, 50, 100, 500, 1000, 10000, 500000000]
    counts_gas, edges_gas = np.histogram(M_no_offshore['drygas'][:, 1], bins=edges_gas_set)

    bin_ave_gas = np.zeros(len(counts_gas))
    bin_sum_gas = np.zeros(len(counts_gas))
    bin_sum_oilwg = np.zeros(len(counts_gas))

    for i in range(1, len(edges_gas_set)):
        bin_mask = (M_no_offshore['drygas'][:, 1] > edges_gas_set[i - 1]) & (
                    M_no_offshore['drygas'][:, 1] <= edges_gas_set[i])
        bin_ave_gas[i - 1] = np.mean(M_no_offshore['drygas'][bin_mask, 1])
        bin_sum_gas[i - 1] = np.sum(M_no_offshore['drygas'][bin_mask, 1])
        bin_sum_oilwg[i - 1] = np.sum(M_no_offshore['drygas'][bin_mask, 0])

    OPGEE_bin = {
        'gasdry': np.column_stack([counts_gas, bin_ave_gas, bin_sum_gas, bin_sum_oilwg])
    }

    # BINNING - ASSOC GAS WELLS
    counts_gas_assoc, edges_gas_assoc = np.histogram(M_no_offshore['gaswoil'][:, 1], bins=edges_gas_set)

    bin_ave_gas_assoc = np.zeros(len(counts_gas_assoc))
    bin_sum_gas_assoc = np.zeros(len(counts_gas_assoc))
    bin_sum_oilwg_assoc = np.zeros(len(counts_gas_assoc))

    for i in range(1, len(edges_gas_set)):
        bin_mask_assoc = (M_no_offshore['gaswoil'][:, 1] > edges_gas_set[i - 1]) & (
                    M_no_offshore['gaswoil'][:, 1] <= edges_gas_set[i])
        bin_ave_gas_assoc[i - 1] = np.mean(M_no_offshore['gaswoil'][bin_mask_assoc, 1])
        bin_sum_gas_assoc[i - 1] = np.sum(M_no_offshore['gaswoil'][bin_mask_assoc, 1])
        bin_sum_oilwg_assoc[i - 1] = np.sum(M_no_offshore['gaswoil'][bin_mask_assoc, 0])

    OPGEE_bin['gasassoc'] = np.column_stack(
        [counts_gas_assoc, bin_ave_gas_assoc, bin_sum_gas_assoc, bin_sum_oilwg_assoc])

    # BINNING - OIL WELLS
    edges_oil_set = [0, 1, 5, 10, 20, 50, 100, 500, 1000, 10000, 1500000]
    counts_oil, edges_oil = np.histogram(M_no_offshore['oilwgas'][:, 1], bins=edges_oil_set)

    bin_ave_gas_oil = np.zeros(len(counts_oil))
    bin_sum_gas_oil = np.zeros(len(counts_oil))
    bin_sum_oilwg_oil = np.zeros(len(counts_oil))

    for i in range(1, len(edges_oil_set)):
        bin_mask_oil = (M_no_offshore['oilwgas'][:, 1] > edges_oil_set[i - 1]) & (
                    M_no_offshore['oilwgas'][:, 1] <= edges_oil_set[i])
        bin_ave_gas_oil[i - 1] = np.mean(M_no_offshore['oilwgas'][bin_mask_oil, 1])
        bin_sum_gas_oil[i - 1] = np.sum(M_no_offshore['oilwgas'][bin_mask_oil, 1])
        bin_sum_oilwg_oil[i - 1] = np.sum(M_no_offshore['oilwgas'][bin_mask_oil, 0])

    OPGEE_bin['oilwgas'] = np.column_stack([counts_oil, bin_ave_gas_oil, bin_sum_gas_oil, bin_sum_oilwg_oil])

    # BINNING - WET WELLS (The Missing Oil Bin Assignment)
    edges_oil_set_wet = [0, 0.5, 1, 10, 1500000]
    counts_oil_wet, edges_oil_wet = np.histogram(M_no_offshore['oil'][:, 1], bins=edges_oil_set_wet)

    bin_ave_gas_wet = np.zeros(len(counts_oil_wet))
    bin_sum_gas_wet = np.zeros(len(counts_oil_wet))
    bin_sum_oilwg_wet = np.zeros(len(counts_oil_wet))

    for i in range(1, len(edges_oil_set_wet)):
        bin_mask_wet = (M_no_offshore['oil'][:, 1] > edges_oil_set_wet[i - 1]) & (
                    M_no_offshore['oil'][:, 1] <= edges_oil_set_wet[i])
        bin_ave_gas_wet[i - 1] = np.mean(M_no_offshore['oil'][bin_mask_wet, 1])
        bin_sum_gas_wet[i - 1] = np.sum(M_no_offshore['oil'][bin_mask_wet, 1])
        bin_sum_oilwg_wet[i - 1] = np.sum(M_no_offshore['oil'][bin_mask_wet, 0])

    OPGEE_bin['oil'] = np.column_stack([counts_oil_wet, bin_ave_gas_wet, bin_sum_gas_wet, bin_sum_oilwg_wet])

    # BINNING - ALL WELLS
    M_all = np.vstack([M_no_offshore['oilall'], M_no_offshore['gasall']])
    edges_set = [0, 1, 5, 10, 20, 50, 100, 500, 1000, 10000, 1500000]
    counts, edges = np.histogram(M_all[:, 1], bins=edges_set)

    bin_ave_gas_all = np.zeros(len(counts))
    bin_sum_gas_all = np.zeros(len(counts))
    bin_sum_oilwg_all = np.zeros(len(counts))

    for i in range(1, len(edges_set)):
        bin_mask_all = (M_all[:, 1] > edges_set[i - 1]) & (M_all[:, 1] <= edges_set[i])
        bin_ave_gas_all[i - 1] = np.mean(M_all[bin_mask_all, 1])
        bin_sum_gas_all[i - 1] = np.sum(M_all[bin_mask_all, 1])
        bin_sum_oilwg_all[i - 1] = np.sum(M_all[bin_mask_all, 0])

    OPGEE_bin['all'] = np.column_stack([counts, bin_ave_gas_all, bin_sum_gas_all, bin_sum_oilwg_all])


    output_dir = 'Outputs'
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)

    if Basin_Select == -1:
        file_name = 'plot_dat_US.xlsx'
    else:
        file_name = f'plot_dat_{Basin_Index[Basin_Select]}out.xlsx'

    return plot_dat, data_tab, OPGEE_bin
